src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,870 @@
+  // that ~ is a strictly decreasing function, if new_hi is the largest value
+  // not larger than hi that satisfies {bits._zeros, bits._ones}, then ~new_hi
+  // is the smallest value not smaller than ~hi that satisfies
+  // {bits._ones, bits._zeros}
+  U new_hi = ~adjust_lo(~bounds._hi, {bits._ones, bits._zeros});
+  if (new_hi > bounds._hi) {
+    return AdjustResult<RangeInt<U>>::make_empty();
+  }
+
+  bool progress = (new_lo != bounds._lo) || (new_hi != bounds._hi);
+  bool present = new_lo <= new_hi;
+  return {progress, present, {new_lo, new_hi}};
+}
+
+// Try to tighten the known bit constraints from the bound information by
+// extracting the common prefix of lo and hi and combining with the current
+// bit constraints
+// E.g: lo = 010011
+//      hi = 010100,
+// then all values in [lo, hi] would be
+//           010***
+template <class U>
+static AdjustResult<KnownBits<U>>
+adjust_bits_from_bounds(const KnownBits<U>& bits, const RangeInt<U>& bounds) {
+  // Find the mask to filter the common prefix, all values between bounds._lo
+  // and bounds._hi should share this common prefix in terms of bits
+  U mismatch = bounds._lo ^ bounds._hi;
+  // Find the first mismatch, all bits before it is the same in bounds._lo and
+  // bounds._hi
+  U match_mask = mismatch == 0 ? std::numeric_limits<U>::max()
+                               : ~(std::numeric_limits<U>::max() >> count_leading_zeros(mismatch));
+  // match_mask & bounds._lo is the common prefix, extract zeros and ones from
+  // it
+  U new_zeros = bits._zeros | (match_mask & ~bounds._lo);
+  U new_ones = bits._ones | (match_mask & bounds._lo);
+  bool progress = (new_zeros != bits._zeros) || (new_ones != bits._ones);
+  bool present = ((new_zeros & new_ones) == 0);
+  return {progress, present, {new_zeros, new_ones}};
+}
+
+// Try to tighten both the bounds and the bits at the same time
+// Iteratively tighten 1 using the other until no progress is made.
+// This function converges because at each iteration, some bits that are
+// unknown is made known. As there are at most 64 bits, the number of
+// iterations should not be larger than 64
+template <class U>
+static SimpleCanonicalResult<U>
+canonicalize_constraints_simple(const RangeInt<U>& bounds, const KnownBits<U>& bits) {
+  AdjustResult<KnownBits<U>> nbits = adjust_bits_from_bounds(bits, bounds);
+  if (!nbits._is_result_consistent) {
+    return SimpleCanonicalResult<U>::make_empty();
+  }
+  AdjustResult<RangeInt<U>> nbounds{true, true, bounds};
+  // Since bits are derived from bounds in the previous iteration and vice
+  // versa, if one does not show progress, the other will also not show
+  // progress, so we terminate early
+  while (true) {
+    nbounds = adjust_bounds_from_bits(nbounds._result, nbits._result);
+    if (!nbounds._progress || !nbounds._is_result_consistent) {
+      return {nbounds._is_result_consistent, nbounds._result, nbits._result};
```

Comment:
That sounds a little scary to me:
It sounds like we could have inconsistent results: `!nbounds._is_result_consistent` and then simply return the `nbounds._result`. Is that ok, or just confusing wording?
I think again here `empty` would be our friend: returning an empty result is more sane than returning an inconsistent one :sweat_smile:

Reply:
If the `_present` flag is `false` then it does not matter what is in the `result`s.